node/core/provisioner/src/disputes/mod.rs

```diff
@@ -0,0 +1,403 @@
+}
+
+struct PartitionedDisputes {
+	// Active disputes without enough onchain votes to conclude. These
+	// will be sent to the Runtime with FIRST priority.
+	pub cant_conclude_onchain: Vec<(SessionIndex, CandidateHash)>,
+	// Active disputes completely unknown onchain. Will be sent to the
+	// Runtime with SECOND priority.
+	pub unknown_onchain: Vec<(SessionIndex, CandidateHash)>,
+	// Active disputes with enough votes to conclude onchain. These will
+	// conclude (or have already concluded) even if we don't send any votes.
+	// Anyway - they will be sent to the Runtime with THIRD priority.
+	pub can_conclude_onchain: Vec<(SessionIndex, CandidateHash)>,
+	// Concluded disputes which are known onchain. These are not
+	// interesting and won't be sent to the Runtime.
+	pub concluded_known_onchain: Vec<(SessionIndex, CandidateHash)>,
+	// Concluded local disputes which are completely unknown for the Runtime.
+	// Hopefully this should never happen. Will be sent to the Runtime with FOURTH priority.
+	pub concluded_unknown_onchain: Vec<(SessionIndex, CandidateHash)>,
+}
+
+impl PartitionedDisputes {
+	fn new() -> PartitionedDisputes {
+		PartitionedDisputes {
+			cant_conclude_onchain: Vec::new(),
+			unknown_onchain: Vec::new(),
+			can_conclude_onchain: Vec::new(),
+			concluded_known_onchain: Vec::new(),
+			concluded_unknown_onchain: Vec::new(),
+		}
+	}
+
+	fn into_iter(self) -> impl Iterator<Item = (SessionIndex, CandidateHash)> {
+		self.cant_conclude_onchain
+			.into_iter()
+			.chain(self.unknown_onchain.into_iter())
+			.chain(self.can_conclude_onchain.into_iter())
+			// .chain(self.concluded_known_onchain.into_iter()) -> dropped on purpose
+			.chain(self.concluded_unknown_onchain.into_iter())
+	}
+}
+
+fn partition_recent_disputes(
+	recent: Vec<(SessionIndex, CandidateHash, DisputeStatus)>,
+	onchain: &HashMap<(u32, CandidateHash), DisputeState>,
+) -> PartitionedDisputes {
+	let mut partitioned = PartitionedDisputes::new();
+
+	// Drop any duplicates
+	let unique_recent = recent
+		.into_iter()
+		.map(|(session_index, candidate_hash, dispute_state)| {
+			((session_index, candidate_hash), dispute_state)
+		})
+		.collect::<HashMap<_, _>>();
+
+	// Split ACTIVE from CONCLUDED disputes
+	let (active, concluded): (
+		Vec<(SessionIndex, CandidateHash, DisputeStatus)>,
+		Vec<(SessionIndex, CandidateHash, DisputeStatus)>,
+	) = unique_recent
+		.into_iter()
+		.map(|((session_index, candidate_hash), dispute_state)| {
+			(session_index, candidate_hash, dispute_state)
+		})
+		.partition(|(_, _, status)| *status == DisputeStatus::Active);
+
+	// Split ACTIVE in three groups...
+	for (session_index, candidate_hash, _) in active {
+		match onchain.get(&(session_index, candidate_hash)) {
+			Some(d) => {
+				// handle double voting here - if a validator has voted both `for` and `against` at the same time for a dispute - ignore its vote
+				let double_votes =
+					(d.validators_for.clone() & d.validators_against.clone()).count_ones() * 2;
```

Comment:
Is there a way to AND two `BitVec`-s without cloning them? I'm missing something quite obvious so hints are welcome.

Reply:
not sure, but the second clone (d.validators_against) should not be necessary
btw, why not do xor (^) to exclude double votes instead of computing and subtracting 2 * them?

Reply:
Yes, second clone is really not necessary.

No reason for not doing an XOR... thanks @ordian